### ISSUE TITLE
🐛 Word-level diff highlighting seems limited

### ISSUE THREAD

**Issue by DamienCassou** [dandavison/delta#1129]
`1.js`

```js
function foo() {
    return bar();
}
```

`2.js`

```js
function foo() {
    return (
        bar() &&
        baz()
    );
}
```

## Expected (generated by Emacs' `diff-mode`)

![2022-07-19-153130](https://user-images.githubusercontent.com/217543/179762841-540f18cb-38eb-4d37-8622-a77967f632fb.png)

- "&&" and "baz()" are highlighted to show they were added
- the rest is not highlighted

## Actual

```session
$ delta --light 1.js 2.js 
```

results in:

![2022-07-19-153144](https://user-images.githubusercontent.com/217543/179762848-50cd345f-08d8-4f75-a22f-123db3e6f2fe.png)

- "&&" and "baz()" have the same highlighting as "bar()"

**Comment by timsu92** [comment]
Me too.

Result of compare in GitHub.com:
![image](https://user-images.githubusercontent.com/33785401/187716794-5b9bf540-0f1c-4629-8806-180441911344.png)
Result of delta
![image](https://user-images.githubusercontent.com/33785401/187716974-d60a285e-eb9a-4df1-95c4-22ed32f6bfa9.png)

**Comment by dandavison** [comment]
Hi @DamienCassou, @timsu92. I think you can get what you want if you experiment with `plus-emph-style` and `plus-non-emph-style`. E.g.

```
delta --plus-style='syntax #a0efa0' --plus-non-emph-style='syntax #d0ffd0' a.js b.js
```

<img width="733" alt="image" src="https://user-images.githubusercontent.com/52205/187732759-3e43635b-ed90-4f78-903e-2153790059bc.png">

**Comment by timsu92** [comment]
Sorry, but no😭

I meant the content it considers to be emphasized, not the color.
From [my comment](https://github.com/dandavison/delta/issues/1129#issuecomment-1233086756), we can see that `silent!` is considered to be removed, which is correct. However, ` :let` (with a preceding space) is considered to be added.
The first point is that it included an extra space. The second is that the space between `:let` and `b:coc_diagnostic...` should be included as well.

**Comment by dandavison** [comment]
@timsu92 in addition to screenshots, can you post the raw diff input to delta as text please, so that others can experiment with piping it to delta. It may be hard to exactly recreate particular expectations regarding highlighting of whitespace, although there are some settings you could play with such as `word-diff-regex`.

**Comment by DamienCassou** [comment]
@dandavison I'm sorry but I don't understand your proposal. The screenshot you attach seems to contain the exact same problem as the one I describe in the post: your screenshot makes absolutely no difference between `bar` and `baz` even though one was there in the original text and one wasn't. Said differently, I would expect `baz` to be highlighted (as it's new) but not `bar` (as it hasn't changed).

**Comment by dandavison** [comment]
Hi @DamienCassou, Emacs' algorithm seems to have produced quite nice results on your example. Delta's algorithm is documented here: [ARCHITECTURE.md#handling-diff-hunk-lines](https://github.com/dandavison/delta/blob/master/ARCHITECTURE.md#handling-diff-hunk-lines). So, as you can see, the first important thing to notice is that delta works by first "pairing up" lines to decide which ones it considers to be homologs (the "same" line, just changed).

In your example, delta pairs up the two lines containing the word `return`. Then, it sees two newly added lines: one containing `bar` and one containing `baz`.

What you would like it to do, is instead consider `    return (` to be a newly inserted line, and  pair up the line

```
    return bar();
```
with
```
        bar() &&
```

It doesn't do that, because it computes a distance metric between the lines, and finds `return bar()` to be more similar to `return (` than to `bar() &&`.

We can tweak your example to demonstrate delta doing something like you want (but obviously that's not a solution for you :)  :

1.js
```
function foo() {
    return bar() bar();
}
```

2.js
```
function foo() {
    RETURN (
        bar() bar() &&
        baz()
    );
}
```

Notice that here we've

1. Made the "return" lines dissimilar
2. Increased the proportion of matching in the `bar` line by making them longer.

<img width="854" alt="image" src="https://user-images.githubusercontent.com/52205/188130168-abbd4d22-a40e-4bfc-ae4e-c08b5f5b4e75.png">

If you're interested in working more on the within-line diff algorithm that would be great. The possibilities definitely haven't been explored thoroughly yet, although the current algorithm (pair lines by similarity, tokenize, and run a levenshtein-like edit inference algorithm) works quite well and people haven't really demanded tweaks over the last couple of years.

**Comment by dandavison** [comment]
@timsu92 I agree that delta could do a better job / there are bugs with the way edits are inferred/highlighted in the whitespace. If you would like to open a separate issue for that that would be great (there might already be one though).

**Comment by timsu92** [comment]
@dandavison OK, submitted as #1181. Hope a better delta coming soon~

**Comment by cben** [comment]
> So, as you can see, the first important thing to notice is that delta works by first "pairing up" lines to decide which ones it considers to be homologs (the "same" line, just changed).

Interesting.  While sophisticated, it sounds like the idea pairing up lines is inherently limited in handling cases where lines get split up / joined? :thinking: 

kosaki/colordiff.py is a useful strawman because it's extremely short & understandable — and it does better here, because it's feeding _whole subhunk vs. whole subhunk_ (accumulated in self.minus_buf / self.plus_buf, complete with newlines) into token-level diffing:
https://github.com/kosaki/colordiff.py/blob/cdd1c1181f6324621272f4ee5ba35081a21a568d/colordiff.py#L83-L99

![colordiff.py screenshot](https://github.com/dandavison/delta/assets/273688/63fa6f6c-a4e9-46b4-9d8e-a72542a88cb1)

<details>

<summary>(tokenization is not essential here)</summary>

working on tokens instead of characters is a simplistic tradeoff to highlight whole words instead of mid-word character changes (which tend to be confusing if unrelated words get "paired up"); the specific tokens it compared in this example are:  
`['-', '    ', 'return', ' ', 'bar', '(', ')', ';', '\n']`  
vs.  
`['+', '    ', 'return', ' ', '(', '\n', '+', '        ', 'bar', '(', ')', ' ', '&', '&', '\n', '+', '        ', 'baz', '(', ')', '\n', '+', '    ', ')', ';', '\n']`

</details>

The essential point is that treating whole subhunk as one sequence with newlines can deal with newline insertion/deletion as a minor internal diff, without throwing off matching of other text.

I think meld works similarly, appears to feed whole (sub)hunks into character-level diffing:

![meld screenshot](https://github.com/dandavison/delta/assets/273688/2f815a9c-205f-43bf-b177-b788973a39b0)

----

Another meld example dealing well with multiple line joining/splitting/indentation changes (though I had to lengthen `bar()` to `barbar()` to not get swallowed in adjacent changes):
![meld screenshot](https://github.com/dandavison/delta/assets/273688/04aab7b5-2ec1-49a3-ad9a-60779058c31d)

**Comment by dbaarda** [comment]
The pairing up of lines algorithm also has some pretty degenerative behaviour in side-by-side mode too.

For each line in a removed hunk it just looks for the next line added in the corresponding added hunk up to `-line-buffer-size` ahead and pairs it if the best matching line has a "distance" score less than `--max-line-distance`. It only considers it to be a deleted line if it didn't find a match close enough. The problem is, a single genuinely deleted line might happen to match close enough to some other arbitrary added line up to `-line-buffer-size` ahead, and this then throws out the line matching for all the subsequent lines.

Playing with  `--max-line-distance` and `--line-buffer-size` I can't seem to make it work nicely, and the only way I can make it line-match in a more optimal way is to flip the diff of the files so that there are less deleted lines at the start of hunks, and instead there are inserted lines at the start of the hunks. The problem is in typical diff targets there are commonly repeated line-patterns that almost match, so trying to set `--max-line-distance` strict enough to reject pairing inserted lines typically results in setting it so strict it can't match genuinely modified lines either. For some reason trying to fix the problem by setting `--line-buffer-size` shorter to prevent it from trying to find matching lines too far ahead completely breaks line-pairing for hunks that are larger than `--line-buffer-size`.

The other option to play with is `--word-diff-regex` which changes what it considers to be a "word", but I've also not had much success playing with that either.

An optimal line matching algorithm would pair lines to give the lowest total line distance for all the lines in the whole deleted/inserted hunk. There are probably simple heuristics that can get close to this that would be much better than the current algorithm.

**Comment by dbaarda** [comment]
Some thoughts on different ways to improve this;

1. A simple fix that might make it much better is instead of always pairing each next-removed-line with the best-and-good-enough-match next-inserted-line, also try pairing the next-inserted-line with the best-and-good-enough-match next-removed-line, and pick the match that has the smallest look-ahead. Or maybe just best distance score.
2. Somehow incorporate the number of lines skipped over into the distance metric, so the distance score gets worse for lines that are further away. This could be combined with option 1.
3. Use `git diff --word-diff=porcelain --diff-algorithm=minimal` to compare hunks for identifying inserted/deleted words instead of the line-pairing algorithm. This seems to do a much better job of identifying inserted/deleted words in matching lines than the pair-matching algorithm.
